### PR TITLE
Add support for softkms and cloudkms uris on requests

### PR DESCRIPTION
**Pull request by maraino**
### Description

This PR allows referring a key using a URIs like "sofkms:/path/to/key.pem" or "cloudkms:projects/p/locations/l/keyRings/k/cryptoKeys/c/cryptoKeyVersions/1".

TBD: 
* Should we use a URI format with key=value pairs? Should we support both?
  * "sofkms:path=/path/to/key.pem"
  * "cloudkms:resource=projects/p/locations/l/keyRings/k/cryptoKeys/c/cryptoKeyVersions/1".
* Should we return the URI in the responses? For both KMSs?

@hslatman What do you think?

### CODE REVIEW

File: kms/softkms/softkms.go [hunk 1]
```diff
@@ -176,3 +177,10 @@ func (k *SoftKMS) CreateDecrypter(req *apiv1.CreateDecrypterRequest) (crypto.Dec
 		return nil, errors.New("failed to load softKMS: please define decryptionKeyPEM or decryptionKey")
 	}
 }
+
+func filename(s string) string {
+	if u, err := uri.ParseWithScheme(string(apiv1.SoftKMS), s); err == nil {
```
Comment (hslatman) [inline kms/softkms/softkms.go:182]: Should this take `Scheme`, and put `Scheme = string(apiv1.SoftKMS)` somewhere for consistency?

**Comment by hslatman** [comment]
In terms of options: I think it's nice if `softkms:/path/to/file` works. It's quick and simple, and looks most like our current paths. I think `file=` could be optional. And if there are other fields, they must have the `key=value` format. So if there's only one property, for `softkms` it'll be the filepath. 

I would say that returning the scheme in the URI in the responses makes sense. That way it's self-contained. If the scheme were to be excluded from the response, it could happen that the value from a response would be used with a different KMS, I guess? With the full URI returned, it could be passed on to the next operation more easily. 

For the TPM KMS I've implemented returning the full URI too. I might add support for not having to specify the `name=<value>` to address the key, and for `tpmkms:<value>` to be enough to identify the key. Will push something soon.

With the above changes, I think we can still support the current `kms` configuration separately from the URIs? Or is your intention to replace that?

**Comment by maraino** [comment]
> With the above changes, I think we can still support the current kms configuration separately from the URIs? Or is your intention to replace that?

I want to support both. I think I will return URIs in `cloudkms` too (all others do that) but probably leave the `softkms` as the filename. Then we can manage that case in generic kms. Sounds that ok?

**Comment by hslatman** [comment]
Sounds good!

File: kms/softkms/softkms.go (hunk 1 above)
Comment (maraino) [reply kms/softkms/softkms.go]: I've added support for all of them in 04cb894

**Comment by maraino** [comment]
@hslatman This should be ready now.

File: kms/cloudkms/cloudkms.go [hunk 2]
```diff
@@ -190,9 +190,12 @@ func (k *CloudKMS) CreateKey(req *apiv1.CreateKeyRequest) (*apiv1.CreateKeyRespo
 
 	var crytoKeyName string
 
+	// resource is the plain Gloud KMS resource name
```
[the remark this answers is not in the document]
Comment (maraino) [reply kms/cloudkms/cloudkms.go]: Fixed with ecd8cc5

File: kms/cloudkms/cloudkms.go [hunk 3]
```diff
@@ -357,3 +363,19 @@ func parent(name string) (string, string) {
...
 	}
 }
+
+// resourceName returns the resource name in the given string. The resource name
+// can be the same string, the value of the resource field or the encoded opaque
+// data:
+//   - projects/id/locations/global/keyRings/ring/cryptoKeys/root-key/cryptoKeyVersions/1
+//   - cloudkms:resource=projects/id/locations/global/keyRings/ring/cryptoKeys/root-key/cryptoKeyVersions/1
+//   - cloudkms:projects/id/locations/global/keyRings/ring/cryptoKeys/root-key/cryptoKeyVersions/1
+func resourceName(name string) string {
```
[the remark this answers is not in the document]
Comment (maraino) [reply kms/cloudkms/cloudkms.go:373]: The main reason for this method is to support backward compatibility. The `CreateKey` method now returns the URI that you should use, in this case, `cloudkms:projects/id/...`

**Review by hslatman — APPROVED** [review]

**Pull request merged**